### ISSUE TITLE
Builtins writing to stdout don't abort when stdout has closed

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#9499]
Using `head` to terminate long-running command after x lines is a common idiom. Here, it's used to stop reading a terabyte log file after sufficient matches are found via `grep`:

```
> tac /hddpool/nginx.access.log | grep -i keep-alive | head -n8
<matches elided> # to stdout, from `grep`
/hddpool/nginx.access.log: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" } # to stderr, from `tac`
```

but with `string match` instead of grep:

```
> tac /hddpool/nginx.access.log | string match -ei keep-alive | head -n8
<matches elided> # to stdout, from `grep`
# script hangs here w/ `tac` running but `head` aborted (currently a zombie)
```

When stdout is closed, builtins need to terminate and then exit (closing their `stdin` in the process).

I already patched all the builtin write functions to return success/failure in 3.6.0, but we're going to need to start checking the return code in each builtin as it writes to stdout so we can abort early.

**Comment by faho** [comment]
For `string match` specifically, variable importing makes this more complex.

We do keep the *first* matched capturing group (without `--all`), but if we don't yet have one it seems correct to keep on trying. With `--all` we might need to exhaust stdin anyway.

**Comment by faho** [comment]
Looking more at this, I think the only builtins that really need this are "string" and "path" and maybe "complete --do-complete" - the rest doesn't really produce enough output to be worth it.

Even there, some paths need to exhaust the input first ("path sort").

One decision I would like here is to not return failure just because stdout was closed - if e.g. `path filter` has printed one thing and would return success if that was the last, it's awkward to have it fail just because it can't write more.

**Comment by mqudsi** [comment]
I agree — if I recall correctly, in my previous testing, an abruptly closed stdout fd generally does not result in a change to the exit code of core Unix utilities. (This would only affect `$pipe_status`, in all cases.)

**Comment by IsaacOscar** [comment]
In case it helps anyone, I was investigating my duplicate issue #9499, and found that putting `trap exit SIGPIPE` in your config.fish can help when fish itself is in a pipeline.

E.g. with a `script.fish` like this:
```fish
#!/bin/fish
while read line
   echo $line
end
```
`yes | ./script.fish | false` will only terminate if you've used `trap``

The same thing happens with this:
```fish
#!/bin/fish
cat
```
Which doesn't use any builtins, only fish itself.

However, it doesn't fix problems when a builtin is trying to write to a closed stdout, e.g:
```fish
#!/bin/fish
cat | string match -r .
```
Will hang even with the trap

I haven't been able to get the trap to kill a fish that isn't piped somewhere, so I don't see any downside to having `trap exit SIGPIPE` in my config.fish, so perhaps this should be on by default?
